### PR TITLE
Detect version mismatches of `parity-scale-codec`

### PR DESCRIPTION
**Pull request by cmichi**
Closes https://github.com/paritytech/cargo-contract/issues/232.

CI currently fails because `master` fails (see https://github.com/paritytech/cargo-contract/pull/236 for the fix for that).

@Robbepop @ascjones Are you aware of any other dependencies of a contract which must be compatible with the ink! version of that dependency? `scale-info`?

### CODE REVIEW

**Comment by ascjones** [comment]
Yes the versions of `scale-info` should match

File: src/cmd/build.rs [hunk 1]
```diff
@@ -430,6 +430,23 @@ fn do_optimization(
     Ok(())
 }
 
+/// Asserts that the contract's dependencies are compatible to the ones used in ink!.
+///
+/// This function utilizes `cargo tree`, which takes semver into consideration.
+/// Hence this function only returns an `Err` if it is a proper mismatch in the
+/// major version, and not if the versions mismatch on the minor/patch version.
```
Comment (ascjones) [inline src/cmd/build.rs:437]: Hm at the moment `scale-info` is < `1.0` so breaking changes might come in minor version bumps.

Comment (cmichi) [reply src/cmd/build.rs:437]: Well then the versioning for `scale-info` is broken. Semver is pretty clear about this:

>MAJOR version when you make incompatible API changes,
>MINOR version when you add functionality in a backwards compatible manner, and

Comment (cmichi) [reply src/cmd/build.rs:437]: Come to think about it, I know another tool which then doesn't use semver properly 🙈

Comment (cmichi) [reply src/cmd/build.rs:437]: I mean I could hardcode a version mismatch for `scale-info` if the minor version is not the same? That's pretty hacky, but I guess the versioning scheme for `scale-info` won't change anytime soon.

Comment (ascjones) [reply src/cmd/build.rs:437]: > Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable

https://semver.org/#spec-item-4

Comment (ascjones) [reply src/cmd/build.rs:437]: We'll go for `1.0` soon, hopefully in time for `ink! 3.0.0` proper

Comment (cmichi) [reply src/cmd/build.rs:437]: Just noticed that `cargo-tree` behaves correcly with `0.y.z`:
```
$ cargo tree --duplicates -i scale-info
error: There are multiple `scale-info` packages in your project, and the specification `scale-info` is ambiguous.
  scale-info:0.3.0
  scale-info:0.6.0
```
So it already detects version mismatches for `0.`.

**Comment by cmichi** [comment]
That's the current output:
```
$ cargo run -- contract build --manifest-path /tmp/foo/Cargo.toml 
error: There are multiple `scale-info` packages in your project, and the specification `scale-info` is ambiguous.
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:
  scale-info:0.3.0
  scale-info:0.6.0
ERROR: Mismatching versions of `scale-info` were found!
Please ensure that your contract and your ink! dependency use a compatible version of this package.
```
Mismatches of `parith-scale-codec` and `scale-info` are detected.

**Comment by ascjones** [comment]
Looks good but CI failing with compilation error.

**Comment by ascjones** [comment]
CI fix merged to master so just needs merging in here

**Review by ascjones — APPROVED** [review]
Nice! This should save us a lot of time answering this common problem.

**Pull request merged**